feat(rsk): allows running on RSKj (#2)
rsksmart/optimism: docs/rootstock/setup_regtest.md

  </svg>
</div>
  <a href="https://optimism.io"><img alt="Optimism" src="https://raw.githubusercontent.com/ethereum-optimism/brand-kit/main/assets/svg/OPTIMISM-R.svg" width=600></a>
  <br />
  <h3><a href="https://optimism.io">Optimism</a> is <del>Ethereum</del> <ins>Rootstock</ins>, scaled.</h3>
  <br />
</div>

# Guide to setup local L1 and L2 development nodes

You will need **5 TTYs** for running the both **L1** and **L2** nodes.
For the purpose of this guide we will assume:

```shell
export PATH_RSKSMART="$HOME/src/rsksmart/"
```

This is illustrative, but by all means, set it in each TTY to copy-paste from this guide and/or set it to your desired location.

#### TTY 1

1. clone op-geth to serve as the sequencer client and build it

    ```shell
    cd $PATH_RSKSMART  # if not there already
    git clone git@github.com:rsksmart/op-geth.git op-sequencer
    cd op-sequencer

>     git swich rsk/initial_attempt

Reviewer: We may need to change it to point to rsk/poc-v0 or just set it as default branch

Author: done

    make geth
    ```

2. clone this repo and switch to this branch:

    ```shell
    cd $PATH_RSKSMART  # if not there already
    git clone git@github.com:rsksmart/optimism.git op-rootstock
    cd op-rootstock

>     git switch rsk/juraj/initial

Reviewer: We may need to change it to point to rsk/poc-v0 or just set it as default branch

Author: done

<div align="center">
  <div>

>   <svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="100%" height="100px" viewBox="39.3 0 100 29.7" style="background-color: rgb(255, 255, 255);">

Reviewer: The images doesn't appear as it should 😓
But it was a nice idea, probably we need to ask Ahmad to provide us with a name/logo.
Please have a look at [this](https://github.com/rsksmart/optimism/tree/rsk/juraj/setup)

Author: yea I noticed, didn't want to waste more time on it tho. Could you please check the link? It seems to just point to this branch, not sure if that's what you wanted...?

Reviewer: Yup, that's what I wanted, to highlight that the first line of the image (ROOTSTOCK) isn't visible

Author: will remove

Author: done

  </svg>
</div>
  <a href="https://optimism.io"><img alt="Optimism" src="https://raw.githubusercontent.com/ethereum-optimism/brand-kit/main/assets/svg/OPTIMISM-R.svg" width=600></a>
  <br />
  <h3><a href="https://optimism.io">Optimism</a> is <del>Ethereum</del> <ins>Rootstock</ins>, scaled.</h3>
  <br />
</div>

# Guide to setup local L1 and L2 development nodes

> You will need **5 TTYs** for running the both **L1** and **L2** nodes.

Reviewer: ```suggestion
You will need **6 TTYs** for running the both **L1** and **L2** nodes.
```

Author: 5 to run the nodes.
6th is for testing ;)

For the purpose of this guide we will assume:

```shell
export PATH_RSKSMART="$HOME/src/rsksmart/"
```

This is illustrative, but by all means, set it in each TTY to copy-paste from this guide and/or set it to your desired location.

> #### TTY 1

Reviewer: Should we include a step 0 to verify that the requirements are met?

Author: done by adding link to optimism documentation

1. clone op-geth to serve as the sequencer client and build it

    ```shell
    cd $PATH_RSKSMART  # if not there already
    git clone git@github.com:rsksmart/op-geth.git
    cd op-geth
    git swich rsk/poc-v0
    make geth
    ```

2. clone this repo and switch to this branch:

    ```shell
    cd $PATH_RSKSMART  # if not there already
    git clone git@github.com:rsksmart/optimism.git
    cd optimism
    git switch rsk/poc-v0
    ```

3. make a copy of the `.envrc.example` and configure it according to:

    [fill-out-environment-variables](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup#fill-out-environment-variables) optimism documentation

    ```shell
    cd $PATH_RSKSMART/optimism # if not there already
    make rsk-env # this will also create the configuration in the future
    ```

      > [!NOTE]

>       > These congfs shell be automated in the near future (//TODO:).

Reviewer: ```suggestion
      > These configs shall be automated in the near future (// TODO:).
```

Author: done

>       - additionally, to make the use of the rsk make targets configure following env vars:

Reviewer: ```suggestion
      - additionally, to make the use of the rsk make targets configure the following env vars:
```

Author: done

    ```shell
    # path to optimism
    export PATH_OPSTACK="$PATH_RSKSMART/optimism"

    # path to bedrock contracts
    export PATH_CONTRACTS="$PATH_OPSTACK/packages/contracts-bedrock"

    # paths to clients
    export PATH_CLIENT_CONSENSUS="$PATH_OPSTACK/op-node"
    export PATH_CLIENT_SEQUENCING="$PATH_RSKSMART/op-geth"

    # regtest 33
    export L1_CHAIN_ID=33
    ```

      - also change:
      ```shell
      export DEPLOYMENT_CONTEXT=regtest
      ```

    > [!IMPORTANT]
    > run `direnv allow` to load the environment variables after change.

4. run your favourite **L1 node**:

    - in docker

    ```shell
    make rsk-regtest-start-log
    ```

    - from jar

    ```shell

>     #TODO:

Reviewer: Instead of a TODO, I'd just add a link to the devportal

Author: done

    ```

#### TTY 2 ([start op geth](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup#start-op-geth))

5. build the internal packages and run sequencer client with

    ```shell

>     make rsk-build

Reviewer: I'd start with `cd $PATH_RSKSMART/optimism`, because it's the first time we use TTY 2

Author: done

>     make rsk-sequencer-fresh # this consists of rsk-wallets rsk-config rsk-create2 rsk-deploy rsk-deploy-sync rsk-create-genesis rsk-jwt rsk-sequencer-init rsk-sequencer-run

Reviewer: I'd remove the comment from here, because if we change the `rsk-sequencer-fresh` target, we need to remember that we have to change the comment here.

Author: fair enough

Author: done

6. run consensus client

    ```shell
    make rsk-consensus-run
    ```

#### TTY 4 ([start op batcher](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup#start-op-batcher))

7. run batcher client

    ```shell
    make rsk-batcher-run
    ```

#### TTY 5 ([start op proposer](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup#start-op-proposer))

8. run proposer client

    ```shell
    make rsk-proposer-run
    ```

---

| :rocket:    | Liftoff, We Have Liftoff|
|---------------|:----------------------------------------|
||[get hacking](https://docs.optimism.io/builders/chain-operators/hacks/overview)|

#### TTY 6 test setup

deposit some RBTC using sdk

```
cd packages/sdk
npx hardhat deposit-eth --to 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 --amount 100 --network regtest --withdraw false;
```

> [!NOTE]
> the `withdraw false` is necessary at the moment as withdrawal not yet supported on RSK

and to verify the balance

```
cast balance 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 --rpc-url http://localhost:8545

> ```

Reviewer: Not for us, but for externals I'd also mention that if the received balance is greater than 0, then setup is successful

Author: overkill?
